File: spoj/find-the-kth-prime.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

int M = 100000000;
int PRIMES = 5000000;
unsigned char *bools;
int *primes;

void gen_primes(){
    int pi = 0;
    int i, j;
    bools = (unsigned char *)malloc(M / 8);
    memset(bools, (char)0xff, M / 8);
    primes = (int *)malloc(PRIMES * sizeof(int));
    for(i = 2; i < (int)sqrt(M) + 1; i++){
        if(bools[i / 8] >> (i % 8) & 1){
            primes[pi++] = i;
            for(j = i * i; j < M; j += i){
                bools[j / 8] &= ~(1 << (j % 8));
            }
        }
    }
    for(;i < M & pi < PRIMES; i++){
        if(bools[i / 8] >> (i % 8) & 1){
            primes[pi++] = i;
        }
    }
}


int main(){
    gen_primes();
    int t, k;
    scanf("%d\n", &t);
    for(int i = 0; i < t; i++){
        scanf("%d\n", &k);
        printf("%d\n", primes[k - 1]);
    }
    free(bools);
    free(primes);
    return 0;
}

File: spoj/find-the-kth-prime.py
import math

def gen_primes(n):
    bools = [True] * n
    for i in range(2, int(math.sqrt(n)) + 1):
        if bools[i]:
            for j in range(i ** 2, n, i):
                bools[j] = False
    return [i for i in range(2, n) if bools[i]]


def segment_sieve(n):
    root = int(math.sqrt(n)) + 1
    primes = gen_primes(root)
    all_primes = primes[:]
    low = root
    high = root * 2
    while low < n:
        high = min(high, n)
        bools = [True] * (high - low)
        for p in primes:
            i = (low + p - 1) // p * p
            for i in range(i, high, p):
                bools[i - low] = False
        all_primes.extend(low + i for i, b in enumerate(bools) if b)
        low, high = high, high + root
    return all_primes


def main():
    primes = segment_sieve(100000000)
    t = int(input())
    for i in range(t):
        n = int(input())
        print(primes[n - 1])

main()
